File: test/test_full_model_parity.py
#!/usr/bin/env python3
"""
Test full model parity - run a forward pass through the score network
with trained weights and save outputs for comparison with Julia.
"""

import sys
sys.path.insert(0, '/home/claudey/BFlaproteina/la-proteina')
import torch_scatter_compat  # Must be before proteinfoundation imports

import numpy as np
import torch
from pathlib import Path
from omegaconf import OmegaConf

# Load model
from proteinfoundation.nn.local_latents_transformer import LocalLatentsTransformer

# Load checkpoint
ckpt_path = "/home/claudey/BFlaproteina/la-proteina/checkpoints_laproteina/LD1_ucond_notri_512.ckpt"
ckpt = torch.load(ckpt_path, map_location='cpu', weights_only=False)

# Load config
config = OmegaConf.load("/home/claudey/BFlaproteina/la-proteina/configs/nn/local_latents_score_nn_160M.yaml")

# Add cropped_flag_seq to match checkpoint (checkpoint was trained with this feature)
config.feats_seq = list(config.feats_seq) + ["cropped_flag_seq"]

# Create model
model = LocalLatentsTransformer(**OmegaConf.to_container(config), latent_dim=8)

# Load weights (from state_dict with nn. prefix)
state_dict = ckpt['state_dict']
nn_state_dict = {k.replace('nn.', ''): v for k, v in state_dict.items() if k.startswith('nn.')}
model.load_state_dict(nn_state_dict)
model.eval()

print("Loaded model")

# Create deterministic test input
torch.manual_seed(42)
np.random.seed(42)

B, L = 1, 10
t_val = 0.5

# Create input batch (same format as actual inference)
x_t = {
    'bb_ca': torch.randn(B, L, 3) * 0.5,
    'local_latents': torch.randn(B, L, 8) * 0.5
}
t = {
    'bb_ca': torch.full((B,), t_val),
    'local_latents': torch.full((B,), t_val)
}
x_sc = {
    'bb_ca': torch.zeros(B, L, 3),  # No self-conditioning
    'local_latents': torch.zeros(B, L, 8)
}
residue_mask = torch.ones(B, L)

batch = {
    'x_t': x_t,
    't': t,
    'x_sc': x_sc,
    'residue_mask': residue_mask,
    'mask': residue_mask.bool()
}

print(f"Input shapes:")
print(f"  x_t bb_ca: {x_t['bb_ca'].shape}")
print(f"  x_t local_latents: {x_t['local_latents'].shape}")
print(f"  t: {t_val}")

# Forward pass
with torch.no_grad():
    output = model(batch)

# Extract outputs
bb_ca_out = output['bb_ca']['v']  # [B, L, 3]
local_latents_out = output['local_latents']['v']  # [B, L, 8]

print(f"\nOutput shapes:")
print(f"  bb_ca: {bb_ca_out.shape}")
print(f"  local_latents: {local_latents_out.shape}")

print(f"\nOutput values (first 5):")
print(f"  bb_ca[0, 0, :]: {bb_ca_out[0, 0, :]}")
print(f"  local_latents[0, 0, :5]: {local_latents_out[0, 0, :5]}")

# Save for Julia comparison
output_dir = Path('/home/claudey/JuProteina/JuProteina/test')
np.save(output_dir / 'full_model_x_t_bb_ca.npy', x_t['bb_ca'].numpy())
np.save(output_dir / 'full_model_x_t_local_latents.npy', x_t['local_latents'].numpy())
np.save(output_dir / 'full_model_x_sc_bb_ca.npy', x_sc['bb_ca'].numpy())
np.save(output_dir / 'full_model_x_sc_local_latents.npy', x_sc['local_latents'].numpy())
np.save(output_dir / 'full_model_t_val.npy', np.array([t_val]))
np.save(output_dir / 'full_model_mask.npy', residue_mask.numpy())
np.save(output_dir / 'full_model_out_bb_ca.npy', bb_ca_out.numpy())
np.save(output_dir / 'full_model_out_local_latents.npy', local_latents_out.numpy())

print(f"\nSaved test data to {output_dir}")

File: test/test_full_model_parity.jl
#!/usr/bin/env julia
"""
Test full model parity - run a forward pass through the score network
with trained weights and compare outputs with Python reference.
"""

using Pkg
Pkg.activate("/home/claudey/JuProteina/JuProteina")

using JuProteina
using NPZ
using Flux
using LinearAlgebra
using Statistics

# Paths
test_dir = "/home/claudey/JuProteina/JuProteina/test"
weights_path = "/home/claudey/JuProteina/JuProteina/weights/score_network.npz"

# Load Python reference data
println("Loading Python reference data...")
x_t_bb_ca_py = npzread(joinpath(test_dir, "full_model_x_t_bb_ca.npy"))  # [B, L, 3]
x_t_local_latents_py = npzread(joinpath(test_dir, "full_model_x_t_local_latents.npy"))  # [B, L, 8]
x_sc_bb_ca_py = npzread(joinpath(test_dir, "full_model_x_sc_bb_ca.npy"))  # [B, L, 3]
x_sc_local_latents_py = npzread(joinpath(test_dir, "full_model_x_sc_local_latents.npy"))  # [B, L, 8]
t_val = npzread(joinpath(test_dir, "full_model_t_val.npy"))[1]
mask_py = npzread(joinpath(test_dir, "full_model_mask.npy"))  # [B, L]
out_bb_ca_py = npzread(joinpath(test_dir, "full_model_out_bb_ca.npy"))  # [B, L, 3]
out_local_latents_py = npzread(joinpath(test_dir, "full_model_out_local_latents.npy"))  # [B, L, 8]

println("Python output bb_ca[0,0,:]: ", out_bb_ca_py[1, 1, :])
println("Python output local_latents[0,0,:5]: ", out_local_latents_py[1, 1, 1:5])

# Convert to Julia format [D, L, B]
x_t_bb_ca = python_to_julia(x_t_bb_ca_py)
x_t_local_latents = python_to_julia(x_t_local_latents_py)
x_sc_bb_ca = python_to_julia(x_sc_bb_ca_py)
x_sc_local_latents = python_to_julia(x_sc_local_latents_py)
mask = python_to_julia_mask(mask_py)  # [L, B]

B = size(x_t_bb_ca, 3)
L = size(x_t_bb_ca, 2)

println("\nInput shapes (Julia format):")
println("  x_t_bb_ca: ", size(x_t_bb_ca))
println("  x_t_local_latents: ", size(x_t_local_latents))
println("  mask: ", size(mask))
println("  t: ", t_val)

# Create the model
println("\nCreating ScoreNetwork...")
model = ScoreNetwork(
    n_layers=14,
    token_dim=768,
    n_heads=12,
    latent_dim=8,
    dim_cond=256,
    t_emb_dim=256,
    pair_dim=256,
    seq_sep_dim=127,
    xt_pair_dist_dim=30,
    xt_pair_dist_min=0.1f0,
    xt_pair_dist_max=3.0f0,
    x_sc_pair_dist_dim=30,
    x_sc_pair_dist_min=0.1f0,
    x_sc_pair_dist_max=3.0f0
)

# Load weights
println("Loading weights...")
load_score_network_weights!(model, weights_path)

# Create batch dict - model expects nested dicts for x_t, t, x_sc
batch = Dict{Symbol, Any}(
    :x_t => Dict{Symbol, Any}(
        :bb_ca => x_t_bb_ca,
        :local_latents => x_t_local_latents
    ),
    :t => Dict{Symbol, Any}(
        :bb_ca => fill(Float32(t_val), B),
        :local_latents => fill(Float32(t_val), B)
    ),
    :x_sc => Dict{Symbol, Any}(
        :bb_ca => x_sc_bb_ca,
        :local_latents => x_sc_local_latents
    ),
    :mask => mask
)

# Forward pass
println("\nRunning forward pass...")
output = model(batch)

v_bb_ca = output[:bb_ca][:v]  # [3, L, B]
v_local_latents = output[:local_latents][:v]  # [8, L, B]

println("\nJulia output shapes:")
println("  v_bb_ca: ", size(v_bb_ca))
println("  v_local_latents: ", size(v_local_latents))

# Convert to Python format for comparison
v_bb_ca_jl = julia_to_python(v_bb_ca)  # [B, L, 3]
v_local_latents_jl = julia_to_python(v_local_latents)  # [B, L, 8]

println("\nJulia output bb_ca[0,0,:]: ", v_bb_ca_jl[1, 1, :])
println("Julia output local_latents[0,0,:5]: ", v_local_latents_jl[1, 1, 1:5])

# Compare outputs
diff_bb_ca = abs.(v_bb_ca_jl .- out_bb_ca_py)
diff_local_latents = abs.(v_local_latents_jl .- out_local_latents_py)

max_diff_bb_ca = maximum(diff_bb_ca)
max_diff_local_latents = maximum(diff_local_latents)
mean_diff_bb_ca = mean(diff_bb_ca)
mean_diff_local_latents = mean(diff_local_latents)

println("\n" * "="^60)
println("PARITY RESULTS")
println("="^60)
println("bb_ca output:")
println("  Max diff: ", max_diff_bb_ca)
println("  Mean diff: ", mean_diff_bb_ca)
println("  Python: ", out_bb_ca_py[1, 1, :])
println("  Julia:  ", v_bb_ca_jl[1, 1, :])

println("\nlocal_latents output:")
println("  Max diff: ", max_diff_local_latents)
println("  Mean diff: ", mean_diff_local_latents)
println("  Python: ", out_local_latents_py[1, 1, 1:5])
println("  Julia:  ", v_local_latents_jl[1, 1, 1:5])

# Check if parity is achieved
tol = 1e-3
if max_diff_bb_ca < tol && max_diff_local_latents < tol
    println("\n✓ PARITY ACHIEVED! (tol=$tol)")
else
    println("\n✗ PARITY FAILED (tol=$tol)")
    println("  bb_ca max diff: $max_diff_bb_ca")
    println("  local_latents max diff: $max_diff_local_latents")
end
